test/libdnf/preserve_order_map/test_preserve_order_map.cpp

```diff
@@ -0,0 +1,194 @@
+    map2.insert({1, "1"});
+    map2.insert({5, "5"});
+    CPPUNIT_ASSERT(!map2.empty());
+    CPPUNIT_ASSERT_EQUAL(map2.size(), static_cast<size_t>(2));
+    CPPUNIT_ASSERT(map2.count(1) == 1 && map2.count(5) == 1 && map2.count(8) == 0);
+
+    // test copy constructor
+    libdnf::PreserveOrderMap<int, std::string> map3 = map2;
+    CPPUNIT_ASSERT(!map3.empty());
+    CPPUNIT_ASSERT_EQUAL(map3.size(), static_cast<size_t>(2));
+    CPPUNIT_ASSERT(map3.count(1) == 1 && map3.count(5) == 1 && map3.count(8) == 0);
+
+    // test move constructor
+    libdnf::PreserveOrderMap<int, std::string> map4 = std::move(map2);
+    CPPUNIT_ASSERT(!map4.empty());
+    CPPUNIT_ASSERT(map2.empty());
+    CPPUNIT_ASSERT_EQUAL(map4.size(), static_cast<size_t>(2));
+    CPPUNIT_ASSERT(map4.count(1) == 1 && map4.count(5) == 1 && map4.count(8) == 0);
+
+    // test clear()
+    map4.clear();
+    CPPUNIT_ASSERT(map4.empty());
+
+    // test insert
+    auto [it1, inserted1] = map1.insert({1, "1"});
+    CPPUNIT_ASSERT(inserted1);
+    CPPUNIT_ASSERT_EQUAL(it1->first, 1);
+    auto [it2, inserted2] = map1.insert({6, "6"});
+    CPPUNIT_ASSERT(inserted2);
+    CPPUNIT_ASSERT_EQUAL(it2->first, 6);
+    auto [it3, inserted3] = map1.insert({4, "4"});
+    CPPUNIT_ASSERT(inserted3);
+    CPPUNIT_ASSERT_EQUAL(it3->first, 4);
+    auto [it4, inserted4] = map1.insert({6, "6"});
+    CPPUNIT_ASSERT(!inserted4);
+    CPPUNIT_ASSERT_EQUAL(it4->first, 6);
+
+    CPPUNIT_ASSERT_EQUAL(map1.size(), static_cast<size_t>(3));
+    CPPUNIT_ASSERT_EQUAL(map1.count(6), static_cast<size_t>(1));
+
+    // test find
+    auto it = map1.find(4);
+    CPPUNIT_ASSERT_EQUAL(it->first, 4);
+    CPPUNIT_ASSERT_EQUAL(it->second, std::string("4"));
+
+    CPPUNIT_ASSERT(map1.find(5) == map1.end());
+
+    // test erase
+    CPPUNIT_ASSERT_EQUAL(map1.erase(2), static_cast<size_t>(0));
+    CPPUNIT_ASSERT_EQUAL(map1.size(), static_cast<size_t>(3));
+
+    CPPUNIT_ASSERT_EQUAL(map1.erase(6), static_cast<size_t>(1));
+    CPPUNIT_ASSERT_EQUAL(map1.size(), static_cast<size_t>(2));
+    CPPUNIT_ASSERT_EQUAL(map1.count(6), static_cast<size_t>(0));
+
+    // operator[] access existing item
+    CPPUNIT_ASSERT_EQUAL(std::string("4"), map1[4]);
+    CPPUNIT_ASSERT_EQUAL(map1.size(), static_cast<size_t>(2));
+    CPPUNIT_ASSERT_EQUAL(map1.count(4), static_cast<size_t>(1));
+
+    // operator[] modify existing item
+    map1[4] = "4M";
+    CPPUNIT_ASSERT_EQUAL(std::string("4M"), map1[4]);
+    CPPUNIT_ASSERT_EQUAL(map1.size(), static_cast<size_t>(2));
+    CPPUNIT_ASSERT_EQUAL(map1.count(4), static_cast<size_t>(1));
+
+    // operator[] add new item
+    map1[8] = "8New";
+    CPPUNIT_ASSERT_EQUAL(std::string("8New"), map1[8]);
+    CPPUNIT_ASSERT_EQUAL(map1.size(), static_cast<size_t>(3));
+    CPPUNIT_ASSERT_EQUAL(map1.count(8), static_cast<size_t>(1));
+
+    // test at
+    CPPUNIT_ASSERT_EQUAL(std::string("4M"), map1.at(4));
+    CPPUNIT_ASSERT_THROW(map1.at(5), std::out_of_range);
```

Comment:
It's not good to have a huge test that tests everything at once. You've already got the code split to blocks testing the various functionality. In general each block should roughly be one separate test. It can make sense to group some of them together to reduce the repetitive setup, but this is too much...

Reply:
I split code and added more tests.